spec/models/topic_thumbnail_spec.rb

```diff
@@ -2,15 +2,18 @@
 require 'rails_helper'
 
 describe "TopicThumbnail" do
-  let(:upload1) { Fabricate(:image_upload, width: 5000, height: 5000) }
+  let(:upload1) { Fabricate(:image_upload, width: 50, height: 50) }
   let(:topic) { Fabricate(:topic, image_upload: upload1) }
-  let(:upload2) { Fabricate(:image_upload, width: 5000, height: 5000) }
+  let(:upload2) { Fabricate(:image_upload, width: 50, height: 50) }
   let(:topic2) { Fabricate(:topic, image_upload: upload2) }
   let(:upload3) { Fabricate(:upload_no_dimensions) }
   let(:topic3) { Fabricate(:topic, image_upload: upload3) }
 
   before do
     SiteSetting.create_thumbnails = true
+
+    ImageSizer.stubs(:resize).returns([49, 49])
```

Comment:
why do we need this stub?

Reply:
Otherwise, optimized images are not created because of guard

```ruby
class TopicThumbnail < ActiveRecord::Base
  def self.find_or_create_for!(original, max_width: , max_height:)
    ...
    target_width, target_height = ImageSizer.resize(original.width, original.height, { max_width: max_width, max_height: max_height })

    if target_width < original.width && target_height < original.height
      optimized = OptimizedImage.create_for(original, target_width, target_height)
    end
    ...
```
And when the optimized image is missing this spec calculation is failing

```ruby
  # Original + Optimized + 3 theme requests
  expect(thumbnails.length).to eq(5)
```

Reply:
I see, sorry for persisting here :) , but how was it previously working without the stub?

Reply:
All good :) Previously we generated big images 5000x5000

In spec we are calling that method:
```ruby
  def generate_thumbnails!(extra_sizes: [])
    ...
    (Topic.thumbnail_sizes + extra_sizes).each do |dim|
      TopicThumbnail.find_or_create_for!(original, max_width: dim[0], max_height: dim[1])
    end
  end
```

By default `Topic.thumbnail_sizes `returns `[1024, 1024]`

We are passing that `1024` as `max_width` and `max_height` here
```
  def self.find_or_create_for!(original, max_width: , max_height:)
    ...
    target_width, target_height = ImageSizer.resize(original.width, original.height, { max_width: max_width, max_height: max_height })

    if target_width < original.width && target_height < original.height
      optimized = OptimizedImage.create_for(original, target_width, target_height)
    end
```

`ImageSizer` is taking `max_width` and `max_height` into consideration to determine `target_width` and `target_height`

```
module ImageSizer
  # Resize an image to the aspect ratio we want
  def self.resize(width, height, opts = {})
    ...
    max_width = (opts[:max_width] || SiteSetting.max_image_width).to_f
    max_height = (opts[:max_height] || SiteSetting.max_image_height).to_f
    w = width.to_f
    h = height.to_f
    return [w.floor, h.floor] if w <= max_width && h <= max_height
    ...
```

So here is a point, previously, the width and height of generated image were `5000` now it is `50`. So array `[50, 50]` is returned by ImageSizer.

and because we are returning existing image dimensions this condition is not working and optimized images are not generated

```ruby
    if target_width < original.width && target_height < original.height
      optimized = OptimizedImage.create_for(original, target_width, target_height)
    end
```

I had 2 options here:
1. Stub Topic.thumbnail_sizes to return something smaller than my small generated image => this is potentially a better option and will more fit to this spec. 
2. Stub `ImageResizer` to give me something smaller than my 50x50 small image.

Let me change it to stub `thumbnail_sizes` instead of `ImageResizer`